File: tests/test_runtime.py
from dragon_runner.src.runner   import ToolChainRunner, TestResult
from dragon_runner.src.config   import Config
import sys
import os
import subprocess

TEST_DIR            = os.path.dirname(os.path.abspath(__file__))
COMPILE_LIB_SCRIPT  = f"{TEST_DIR}/scripts/test-scripts/compile_lib.py"
LIB_SRC_DIR         = os.path.join(TEST_DIR, "lib/src")
LIB_OUT_DIR         = os.path.join(TEST_DIR, "lib")

def run_tests_for_config(config: Config, expected_result: bool):
    # TODO: move to conftest.py
    assert config.packages is not None

    for exe in config.executables:
        exe.source_env()
        for tc in config.toolchains:
            tc_runner = ToolChainRunner(tc, timeout=3.0)
            for pkg in config.packages:
                for sp in pkg.subpackages:
                    for test in sp.tests:
                        result: TestResult = tc_runner.run(test, exe)
                        result.log()
                        assert result.did_pass == expected_result


def test_gcc_toolchain_success(config_factory, cli_factory):
    assert os.path.exists(COMPILE_LIB_SCRIPT), "missing library compiler script" 

    if sys.platform == "darwin":
        lib = "libfib.dylib"
        config = config_factory("runtimeConfigDarwin.json")
    else:
        lib = "libfib.so"
        config = config_factory("runtimeConfigLinux.json")

    expected_lib=os.path.join(TEST_DIR, f"lib/{lib}")

    if not os.path.exists(expected_lib):
        result = subprocess.run([sys.executable,
                                COMPILE_LIB_SCRIPT,
                                LIB_SRC_DIR,
                                LIB_OUT_DIR], check=True)
        
        assert result.returncode == 0, "shared object compilation failed"
        assert os.path.exists(expected_lib), "failed to create shared object"

    # now shared object exists where the config expects it, so we can run
    os.environ["DRAGON_RUNNER_DEBUG"] = "3"
    run_tests_for_config(config, expected_result=True)

File: dragon-runner-rs/tests/test_runtime.rs
use std::path::Path;
use std::process;

use dragon_runner_rs::config::{load_config, Config};
use dragon_runner_rs::runner::ToolChainRunner;

fn configs_dir() -> std::path::PathBuf {
    let manifest = Path::new(env!("CARGO_MANIFEST_DIR"));
    manifest.parent().unwrap().join("tests").join("configs")
}

fn config_path(name: &str) -> String {
    configs_dir().join(name).to_string_lossy().into_owned()
}

fn tests_dir() -> std::path::PathBuf {
    let manifest = Path::new(env!("CARGO_MANIFEST_DIR"));
    manifest.parent().unwrap().join("tests")
}

fn run_tests_for_config(config: &Config, expected_result: bool) {
    for exe in &config.executables {
        exe.source_env();
        for tc in &config.toolchains {
            let runner = ToolChainRunner::new(tc.clone(), 3.0);
            for pkg in &config.packages {
                for spkg in &pkg.subpackages {
                    for test in &spkg.tests {
                        let result = runner.run(test, exe);
                        assert_eq!(
                            result.did_pass, expected_result,
                            "Test {} expected {} but got {}",
                            test.file,
                            if expected_result { "PASS" } else { "FAIL" },
                            if result.did_pass { "PASS" } else { "FAIL" },
                        );
                    }
                }
            }
        }
    }
}

#[test]
fn test_gcc_toolchain_success() {
    let test_dir = tests_dir();
    let compile_script = test_dir.join("scripts/test-scripts/compile_lib.py");
    let lib_src_dir = test_dir.join("lib/src");
    let lib_out_dir = test_dir.join("lib");

    assert!(compile_script.exists(), "missing compile_lib.py");

    let expected_lib = test_dir.join("lib/libfib.so");
    if !expected_lib.exists() {
        let status = process::Command::new("python3")
            .args([
                compile_script.to_str().unwrap(),
                lib_src_dir.to_str().unwrap(),
                lib_out_dir.to_str().unwrap(),
            ])
            .status()
            .expect("failed to run compile_lib.py");
        assert!(status.success(), "shared object compilation failed");
        assert!(expected_lib.exists(), "failed to create shared object");
    }

    let path = config_path("runtimeConfigLinux.json");
    let config = load_config(&path, None).expect("config should load");
    assert!(
        !config.error_collection.has_errors(),
        "config errors: {}",
        config.error_collection
    );

    run_tests_for_config(&config, true);
}
